src/v/cloud_topics/level_one/frontend_reader/reader.cc

```diff
@@ -0,0 +1,415 @@
+      _io->read_object(extent, abort_source));
+    if (stream_fut.failed()) {
+        vlog(
+          cd_log.error,
+          "Exception opening stream for footer from object {} (pos {} object "
+          "size {}): {}",
+          oid,
+          extent.position,
+          object_size,
+          stream_fut.get_exception());
+        co_return std::unexpected(l1::io::errc::file_io_error);
+    }
+    auto stream_result = stream_fut.get();
+    if (!stream_result.has_value()) {
+        co_return std::unexpected(stream_result.error());
+    }
+
+    auto& stream = stream_result.value();
+
+    auto footer_fut = co_await ss::coroutine::as_future(
+      read_iobuf_exactly(stream, footer_total_size));
+    if (footer_fut.failed()) {
+        vlog(
+          cd_log.error,
+          "Exception reading footer from object {} (pos {} object size {}): {}",
+          oid,
+          extent.position,
+          object_size,
+          footer_fut.get_exception());
+        co_await stream.close();
+        co_return std::unexpected(l1::io::errc::file_io_error);
+    }
+
+    iobuf footer_buf = footer_fut.get();
+    co_await stream.close();
+
+    // Parse the footer - we have the complete footer so this should succeed.
+    auto footer_result = co_await l1::footer::read(std::move(footer_buf));
+
+    if (!std::holds_alternative<l1::footer>(footer_result)) {
+        vlog(
+          cd_log.error,
+          "Failed to parse footer from object {} despite reading complete "
+          "footer (pos {} object size {})",
+          oid,
+          extent.position,
+          object_size);
+        co_return std::unexpected(l1::io::errc::file_io_error);
+    }
+
+    co_return std::get<l1::footer>(std::move(footer_result));
+}
+
+ss::future<> level_one_log_reader_impl::materialize_batches(
+  model::timeout_clock::time_point /*deadline*/) {
+    // Could be EOS because there are no more objects, but
+    // there should never be materialized batches remaining.
+    vassert(
+      _batches.empty(),
+      "Materialize batches called with batches already materialized");
+
+    if (_state == state::end_of_stream) {
+        co_return;
+    }
+
+    vassert(
+      _state == state::ready,
+      "Invalid state to materialize batches: {}",
+      std::to_underlying(_state));
+
+    // I wish I had ADTs to enforce these invariants...
+    vassert(
+      _current_obj.has_value(),
+      "Expected to have current object in ready state");
+
+    size_t start_pos = _current_obj->footer.file_position_before_kafka_offset(
+      _tidp, _config.start_offset);
+    if (start_pos == l1::footer::npos) {
+        // Perhaps this object spans offsets in the metastore but has
+        // no data because of compaction.
+        vlog(
+          cd_log.debug,
+          "No data for {} ({}) in object {}: materializing 0 batches",
+          _ntp,
+          _tidp,
+          _current_obj->oid);
+        _state = state::materialized;
```

Comment:
Don't we need to advance offsets in this case? Otherwise we'll just loop endlessly.